core/src/main/java/org/ethereum/beacon/core/types/Bitlist.java

```diff
@@ -0,0 +1,221 @@
+package org.ethereum.beacon.core.types;
+
+import com.google.common.base.Objects;
+import org.ethereum.beacon.ssz.access.SSZField;
+import org.ethereum.beacon.ssz.access.list.AbstractListAccessor;
+import org.ethereum.beacon.ssz.annotation.SSZSerializable;
+import org.ethereum.beacon.ssz.type.SSZListType;
+import org.ethereum.beacon.ssz.type.SSZType;
+import tech.pegasys.artemis.util.bytes.BytesValue;
+import tech.pegasys.artemis.util.bytes.BytesValues;
+import tech.pegasys.artemis.util.bytes.DelegatingBytesValue;
+import tech.pegasys.artemis.util.bytes.MutableBytesValue;
+import tech.pegasys.artemis.util.uint.UInt64;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@SSZSerializable(listAccessor = Bitlist.BitListAccessor.class)
+public class Bitlist extends DelegatingBytesValue {
+  private final int size;
+  private final long maxSize;
+
+  Bitlist(int size, BytesValue bytes, long maxSize) {
+    super(checkSize(bytes, size));
+    this.size = size;
+    this.maxSize = maxSize;
+  }
+
+  /** Assumes that size info is encoded in bytes\ */
+  public static Bitlist of(BytesValue bytes, long maxSize) {
+    int size = (bytes.size() - 1) * 8;
+    byte lastByte = bytes.get(bytes.size() - 1);
+    int addon = 0;
+    for (int i = 0; i < 8; ++i) {
+      if (((lastByte >> i) & 1) == 1) {
+        addon = i;
+      }
+    }
+    final BytesValue finalBlank;
+    if (addon == 0) {
+      finalBlank = bytes.slice(0, bytes.size() - 1); // last byte was needed only for a size
+    } else {
+      size = size + addon;
+      MutableBytesValue mutableBytes = bytes.mutableCopy();
+      mutableBytes.setBit(size, false);
+      finalBlank = mutableBytes.copy();
+    }
+
+    return new Bitlist(size, finalBlank, maxSize);
+  }
+
+  public static Bitlist of(int size, BytesValue bytes, long maxSize) {
+    return new Bitlist(size, bytes, maxSize);
+  }
+
+  public static Bitlist of(int maxSize) {
+    return new Bitlist(0, BytesValue.EMPTY, maxSize);
+  }
+
+  public static Bitlist of(int size, long bytes, long maxSize) {
+    UInt64 blank = UInt64.valueOf(bytes);
+    if (blank.getUsedBitCount() > size) {
+      throw new IndexOutOfBoundsException(
+          String.format("Input data %s exceeds Bitlist size of %s", bytes, size));
+    }
+
+    int neededBytes = (size + 7) / 8;
+    return new Bitlist(size, blank.toBytesValue().slice(0, neededBytes), maxSize);
+  }
+
+  private static BytesValue checkSize(BytesValue input, int size) {
+    int neededBytes = (size + 7) / 8;
+    if (neededBytes != input.size()) {
+      throw new IllegalArgumentException(
```

Comment:
I think this exception could be reflected in SSZ code as `SSZDeserializeException` for conformance.